Project structure

hershen opened the pull request:
Modify project structure, mostly following http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1204r0.html.
Closes #5.

hershen committed:
Modify project structure

Modify project structure, mostly following http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1204r0.html.
1. Rename header files to .hpp.
2. Move all header, implementation, and test files to hersh sub-directory.
3. The hersh and runTests binaries are in the bin sub-folder.
4. Use #include <> instead of #include "".

hershen committed:
Download the devel branch of Catch2

It appears FetchContent doesn't work with a specific commit hash that's
not at the tip of a branch. So I changed the GIT_TAG to be the Catch2
devel branch.

oorijsonar commented:
I googled a bit and it seems that this is a very uncommon structure. Are you sure that this is how you want it to be?

hershen commented:
Do you see any disadvantages with it, other than that's uncommon?
Have you come across a structure you feel is better?

Another option is to follow [The Pitchfork Layout (PFL)](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs).

oorijsonar commented:
> Do you see any disadvantages with it, other than that's uncommon?
> Have you come across a structure you feel is better?
> 
> Another option is to follow [The Pitchfork Layout (PFL)](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs).

Nope that's the only disadvantage. I like the pitchfork layout better because it seems more standard.

hershen commented:
> > Do you see any disadvantages with it, other than that's uncommon?
> > Have you come across a structure you feel is better?
> > Another option is to follow [The Pitchfork Layout (PFL)](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs).
> 
> Nope that's the only disadvantage. I like the pitchfork layout better because it seems more standard.

OK, I'll update according to pitchfork.

hershen committed:
Change project direcotry structure

The new direcotry sturcutre follows [The Pitchfork Layout
(PFL)](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs).
Headers and sources are in the /src directory.
Binaries are in the /build/bin directory (although this is not specified
in PFL).
External projects (only Catch2 currently) in the /externals/Catch2
directory.

hershen committed:
Make FetchContent follow a specific Catch2 commit

Remove GIT_SHALLOW TRUE as it blocks setting a specific commit for
GIT_TAG.

hershen commented:
> OK, I'll update according to pitchfork.

Done.

oorijsonar commented on CMakeLists.txt:
```diff
@@ -5,34 +5,40 @@ project(hersh)
-set(INCLUDE_DIR include)
-set(TEST_DIR tests)
+set(SOURCE_DIR src)
+set(EXTERNAL_DIR ${PROJECT_SOURCE_DIR}/external)
```
I don't like that we're writing to the sources here but I will fix it when I work on `ExternalProject`.

hershen replied:
Why not? It follows the PFL recommendation, doesn't it?

oorijsonar replied:
Can you post a link to said recommendation?

oorijsonar replied:
In any case, even if that's their recommendation - I prefer not to dirty the repo with things we download when we build the project. It's nicer if we keep that all in `build`.

hershen replied:
[$2.6](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.external) "The external/ directory is reserved for embedding of external projects."
I guess the reasoning is that it makes it consistent that `build` does not include source code, only build results.
But I see your point, and I don't really care.
